core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -624,96 +635,255 @@ public void addPhysicallyWrittenRows(long rows) {
         metrics.tableWriter().addPhysicallyWrittenRows(rows);
     }
 
-    public int attachPartition(long timestamp) {
+    public StatusCode attachPartition(long timestamp) {
         return attachPartition(timestamp, true);
     }
 
-    public int attachPartition(long timestamp, boolean validateDataFiles) {
+    public StatusCode attachPartition(long timestamp, boolean validateDataFiles) {
         // Partitioned table must have a timestamp
         // SQL compiler will check that table is partitioned
         assert metadata.getTimestampIndex() > -1;
 
-        CharSequence timestampCol = metadata.getColumnQuick(metadata.getTimestampIndex()).getName();
         if (txWriter.attachedPartitionsContains(timestamp)) {
-            LOG.info().$("partition is already attached [path=").$(path).$(']').$();
-            return PARTITION_ALREADY_ATTACHED;
+            LOG.info().$("partition is already attached [path=").$(path).I$();
+            // TODO: potentially we can merge with existing data
+            return StatusCode.PARTITION_ALREADY_ATTACHED;
         }
 
         boolean rollbackRename = false;
         try {
-            setPathForPartition(path, partitionBy, timestamp, false);
-            if (!ff.exists(path.$())) {
-                setPathForPartition(other, partitionBy, timestamp, false);
-                other.put(DETACHED_DIR_MARKER);
+            // final name of partition folder after attach
+            setPathForPartition(path.trimTo(rootLen), partitionBy, timestamp, false);
+            path.slash$();
+
+            if (!ff.exists(path)) {
+                setPathForPartition(detachedPath.trimTo(detachedRootLen), partitionBy, timestamp, false);
+                detachedPath.put(ATTACHABLE_DIR_MARKER).slash$();
+                if (ff.exists(detachedPath)) {
+
+                    // check detached _meta and _cv
+                    checkDetachedMetadata(timestamp);
 
-                if (ff.exists(other.$())) {
-                    if (ff.rename(other, path) == Files.FILES_RENAME_OK) {
+                    if (inTransaction()) {
+                        LOG.info().$("committing open transaction before applying attach partition command [table=").$(tableName)
+                                .$(", partition=").$ts(timestamp).I$();
+                        commit();
```

Comment:
What if we modify ColumnVersion records, commit here and fail later on?
Commit should happen before touching Column Versions